crates/core/src/piop/verify.rs

```diff
@@ -126,17 +126,41 @@ where
-	// optimal.
-	let log_batch_size = fold_arities.first().copied().unwrap_or(0);
-	let log_dim = commit_meta.total_vars - log_batch_size;
-
+	let log_dim = commit_meta.total_vars.saturating_sub(arity);
+	let log_batch_size = commit_meta.total_vars.min(arity);
 	let rs_code = ReedSolomonCode::new(log_dim, log_inv_rate)?;
 	let n_test_queries = fri::calculate_n_test_queries::<F, _>(security_bits, &rs_code)?;
+
+	let cap_height = log2_ceil_usize(n_test_queries);
+	let fold_arities = std::iter::repeat_n(
+		arity,
+		(commit_meta.total_vars + log_inv_rate).saturating_sub(cap_height) / arity,
+	)
+	.collect::<Vec<_>>();
+	// here is the down-to-earth explanation of what we're doing: we want the terminal codeword's log-length to be at least as large as the Merkle cap height.
+	// note that `total_vars + log_inv_rate - sum(fold_arities)` is exactly the log-length of the terminal codeword; we want this number to be ≥ cap height.
+	// so fold_arities will repeat `arity` the maximal number of times possible, while maintaining that `total_vars + log_inv_rate - sum(fold_arities) ≥ cap_height` stays true.
+	// this arity-selection strategy can be characterized as: "terminate as late as you can, while maintaining that no Merkle cap is strictly smaller than `cap_height`."
+	// this strategy does attain that property: the Merkle path height of the last non-terminal codeword will equal the log-length of the terminal codeword, which is ≥ cap height by fiat.
+	// moreover, if we terminated later than we are above, then this would stop being true. imagine what would happen if we took the above terminal codeword and continued folding.
+	// in that case, we would Merklize this word, again with the coset-bundling trick; the post-bundling path height would thus be `total_vars + log_inv_rate - sum(fold_arities) - arity`.
+	// but we already agreed (by the maximality of the number of times we subtracted `arity`) that the above number will be < cap_height. in other words, its Merkle cap will be short.
+	// equivalently: this is the latest termination for which the `min` in `optimal_verify_layer` will never trigger; i.e., we will have log2_ceil_usize(n_queries) ≤ tree_depth there.
+	// it can be shown that this strategy beats any strategy which terminates later than it does (in other words, by doing this, we are NOT terminating TOO early!).
+	// this doesn't mean that we should't terminate EVEN earlier (maybe we should). but this approach is conservative and simple; and it's easy to show that you won't lose by doing this.
+
+	// see PR for proof of this fact
+
+	// how should we handle the case `fold_arities = []`, i.e. total_vars + log_inv_rate - cap_height < arity? in that case, we would lose nothing by making the entire thing interleaved,
+	// i.e., setting `log_batch_size := total_vars`, so `terminal_codeword` is an interleaving of the repetition code---so is itself a repetition code!---the prover just sends the message.
```

Comment:
What do you mean the prover just sends the message? They need to encode it with the repetition code. (ie. repeat it)

Reply:
yes, that's what i meant; i can clarify the comment